### PR TITLE
[database] Make season_view and tvshow_view SQL ANSI92-compliant

### PR DESCRIPTION
**Pull request by wildcat2020**
## Description

For season_view: removed Group By clause as it is not needed and it errors out.
For tvshow_view: added full Group By Clause to make SQL ANSI 92 compliant.

## Motivation and Context

This change is related to issue #15172 
"_myvideos113.season_view and tvshow_view not ANSI 92 compliant and therefore broken_"

## How Has This Been Tested?

The SELECT statements were tested via MySQL workbench, however I don't have the means to rebuild kodi from my branch and test so a final test of the view creation statements is needed. Below is description of final test to be made:

- Build new kodi with this commit included.
- Setup in advancesettings.xml video database to open to MySQL 5.7 or later
- Open Kodi and let it create the kodi schema
- Login via MySQL workbench to myvideos113 database as kodi user, and try to query from the below two views:

1. season_view
2. tvshow_view

Query should succeed for both views, returning zero rows.

## Types of change

- [X] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by pkerling** [comment]
Thanks for the PR! I'm no SQL wiz so I hope I don't completely embarrass myself, but wouldn't the `season_view` also work without `GROUP BY`? Grouping by the id does not make sense IMO.

**Comment by wildcat2020** [comment]
Hi, yes you are right, season_view should work without group by as there is no aggregate function in the SELECT clause, therefore I have removed it. However, note that by removing the group by, we are not making the statement 100% equivalent with what it was before, because group by operation includes PROJECTION and SORT operation which removes duplicates. I am not sure how the overall data schema design was done, but assuming this select statement does not return duplicates (I checked for duplicates coming from this view without the group by clause in myvideos113 database and I did not find any), it should be safe and faster (as it removes PROJECTION and hence SORT operation) to remove the group by clause. I hope this makes sense.

However, now that I have looked more into the code into the .cpp I realized that those variables that I saw in the code: %02d do need the values which I saw inside originally in the statement, i.e. for example the below for tvshow_view:
VIDEODB_ID_TV_TITLE, VIDEODB_ID_TV_PLOT, VIDEODB_ID_TV_PREMIERED,
                                     VIDEODB_ID_TV_GENRE, VIDEODB_ID_TV_STUDIOS, VIDEODB_ID_TV_MPAA,
                                     VIDEODB_ID_EPISODE_AIRED, VIDEODB_ID_EPISODE_SEASON)

How can I edit my commit so that I also add those variables back?

**Comment by wildcat2020** [comment]
OK, I think I did it, I added a second commit. Please kinldy check. It should be ok now.

**Comment by wildcat2020** [comment]
Great, I saw it was built so I am downloading 64bit windows one to install and test....will let you know of the test result soon...

**Comment by wildcat2020** [comment]
OK, tested it. season_view was not generated at all, and tvshow_view was generated but still complaining about missing group by. After checking further definition of tvshow_view that is because it includes in SELECT statement tvshowcounts.totalCount, which is defined in tvshowcounts as nullif(count(`myvideos113`.`episode`.`c12`),0)  so it uses aggregate function after all. 
Correcting tvshows_view and checking further why season_view was not created...

**Comment by wildcat2020** [comment]
still not ok...investigaiting...

**Comment by wildcat2020** [comment]
OK, should be corect now after all those changes, but build fails with error:
 $GIT_COMMIT (aca925671e31e99fb1f48d5420f78e41924d2b03) doesn't match currently checked out revision (0308f2c4aae0843bb957c16e3e843f56d019084a)
Any help on how to proceed with build for testing would be appreciated.

File: xbmc/video/VideoDatabase.cpp [hunk 1]
```diff
@@ -394,7 +394,7 @@ void CVideoDatabase::CreateViews()
                                        "        episode.idShow=tvshow.idShow"
                                        "      LEFT JOIN files ON"
                                        "        files.idFile=episode.idFile "
-                                       "    GROUP BY tvshow.idShow");
+                                       "GROUP BY tvshow.idShow");
```
Comment (pkerling) [inline xbmc/video/VideoDatabase.cpp]: Is this change necessary or just a leftover?

Comment (wildcat2020) [reply xbmc/video/VideoDatabase.cpp]: Hi, not actually. But I practically need the final file after all changes as it is here to be the one to be tested: 2a90ad7

Comment (wildcat2020) [reply xbmc/video/VideoDatabase.cpp]: If you can remove all the previous commits somehow and just leave the last image of the file it would be even better :-) I don't know how to do this myself...still new at this GIT system...

Comment (pkerling) [reply xbmc/video/VideoDatabase.cpp]: Hey, we will need you to condense this into one commit before merging anyway :-)

It's called "squashing". You can find a simple explanation here: https://github.com/todotxt/todo.txt-android/wiki/Squash-All-Commits-Related-to-a-Single-Issue-into-a-Single-Commit
In this simple case it would also be possible that you just start fresh (reset the branch to master) and make a new commit from there. You then have to force-push (`git push --force`) because otherwise Git will not allow you to rewrite history.

Don't be afraid to ask if you have further questions.

Comment (wildcat2020) [reply xbmc/video/VideoDatabase.cpp]: Thanks for your response. So please kindly confirm my understanding that I need to issue the below commands from what I read if I use the second option you listed:

1. git reset --hard feature_myvideos113 master        to reset the feature_myvideos113 branch to master
2. git add .     to add the change 
3. git commit -m "final commit"      to create the commit 
3. git push --force  origin feature_myvideos113   to re-push the final videodatabase.cpp file that I have locally in my branch as part of the new commit

Thanks and regards

Comment (pkerling) [reply xbmc/video/VideoDatabase.cpp]: For 1: You first have to checkout the branch with `git checkout feature_myvideos113` if you are not already using it (which I assume). Then just do `git reset master`. If you use `git reset --hard`, it will also reset your files to the master branch and discard all of your changes. Without `--hard`, it just changes the branch pointer and you can then commit the changes again. Rest of the steps seems ok.

Comment (wildcat2020) [reply xbmc/video/VideoDatabase.cpp]: Many thanks for the advice, I'll try those now...

Comment (sopparus) [reply xbmc/video/VideoDatabase.cpp]: this will require a bumb of db version right? is it for v18?

Comment (pkerling) [reply xbmc/video/VideoDatabase.cpp]: Uh why did you not remove the whitespace change here?

Comment (wildcat2020) [reply xbmc/video/VideoDatabase.cpp]: let me check the file again
It seems the space is not there in my file, take a look:

image

Comment (wildcat2020) [reply xbmc/video/VideoDatabase.cpp]: From what I see from other select statements that have group by and I did not change,the identation used is that GROUP BY should follow immediately after the "
This is how I put it in my file to keep the identation consistent.

Comment (pkerling) [reply xbmc/video/VideoDatabase.cpp]: Oh, you are of course absolutely right. Disregard :-)

**Comment by pkerling** [comment]
Thanks a lot for putting your time into this. Code is looking good now, but the commit message needs some work. It is best if you put a short concise and description of the fix into the title and not the issue number, because that is very opaque and does not help when scrolling through a list of commits. I'd suggest `[database] Make season_view and tvshow_view SQL ANSI92-compliant` for example. In the message body, you can describe more in-depth what changes were required to achieve the compatibility and finally say `Fixes #15172`. That magic line will cause GitHub to automatically close the issue once the pull request gets merged.

File: xbmc/video/VideoDatabase.cpp (hunk 1 above)
Comment (wildcat2020) [reply xbmc/video/VideoDatabase.cpp]: No worries, my big fear is whether the season_view will work or not....I hope I used the variables correctly, if I understood it correctly from reading the other select statements...this is why I need to test this...I am not sure why whoever chose to write this file used variables to include in the prepare statement....this is a level of abstraction that I am not used to...

**Comment by pkerling** [comment]
Also as @sopparus suggested it would be good to bump the DB version so Kodi recreates the views that people already might have. It is this line: https://github.com/xbmc/xbmc/blob/52aafcec9abe46833e137b9521640f6dfe69c2b9/xbmc/video/VideoDatabase.cpp#L5444 - just change it to `114`.

File: xbmc/video/VideoDatabase.cpp (hunk 1 above)
Comment (wildcat2020) [reply xbmc/video/VideoDatabase.cpp]: did it again and bumped the database version, please kindly check.

**Comment by pkerling** [comment]
> ..this is why I need to test this...I am not sure why whoever chose to write this file used variables to include in the prepare statement....this is a level of abstraction that I am not used to...

Mostly historic reasons as far as I'm aware. The whole database layer is pending a complete rewrite.

> did it again and bumped the database version, please kindly check.

Version bump is fine. Can you put a message body into the commit message with more details and the "Fixes:" line as I requested please? Right now it seems to be empty.

**Comment by pkerling** [comment]
Oh forgot to mention: SQL statement changes (including the column names) should be fine as far as I can tell.

**Comment by wildcat2020** [comment]
Ok, thanks. I see I can use two times -m command in git commit to add both title and message body, I'll try it out...

**Comment by wildcat2020** [comment]
Ok, done, please check.

**Comment by wildcat2020** [comment]
thanks, a lot. Now that I will do again the change, I think I should also fix one more thing....Although the current changes should make the views created valid, there is still something that is not a good coding practices used in tvshow_view and season_view: who ever wrote the statement, put in SELECT tvshow.* and SELECT seasons.*
If those base tables ever get a new column added, this will break the view again as the new column will automatically appear in the SELECT clause but not in the GROUP BY clause...
Would you like me to remove that as well and add normally the columns of the base tables?

**Comment by pkerling** [comment]
> Would you like me to remove that as well and add normally the columns of the base tables?

Yes, that sounds like a good idea.

**Comment by wildcat2020** [comment]
ok, I'll do this in the next few hours (have to check the tables definitions from MySQL, and will also re-commit with the change in the message body again...have to take out the family now :-)

**Comment by wildcat2020** [comment]
ok did it, please kindly check

File: xbmc/video/VideoDatabase.cpp [hunk 2]
```diff
@@ -443,10 +454,23 @@ void CVideoDatabase::CreateViews()
...
+				     "         tvshow_view.c%02d,"
+				     "         tvshow_view.c%02d,"
+				     "         tvshow_view.c%02d,"
+				     "         tvshow_view.c%02d,"
+				     "         tvshow_view.c%02d ",
                                      VIDEODB_ID_TV_TITLE, VIDEODB_ID_TV_PLOT, VIDEODB_ID_TV_PREMIERED,
                                      VIDEODB_ID_TV_GENRE, VIDEODB_ID_TV_STUDIOS, VIDEODB_ID_TV_MPAA,
-                                     VIDEODB_ID_EPISODE_AIRED, VIDEODB_ID_EPISODE_SEASON);
+                                     VIDEODB_ID_EPISODE_AIRED, VIDEODB_ID_EPISODE_SEASON,
+                                     VIDEODB_ID_TV_TITLE, VIDEODB_ID_TV_PLOT, VIDEODB_ID_TV_PREMIERED,
```
Comment (sualfred) [inline xbmc/video/VideoDatabase.cpp]: duplicate of line 469

File: xbmc/video/VideoDatabase.cpp [hunk 3]
```diff
@@ -443,10 +454,23 @@ void CVideoDatabase::CreateViews()
...
+                                     VIDEODB_ID_TV_GENRE, VIDEODB_ID_TV_STUDIOS, VIDEODB_ID_TV_MPAA);
```
Comment (sualfred) [inline xbmc/video/VideoDatabase.cpp]: dupliciate of line 470

Comment (wildcat2020) [reply xbmc/video/VideoDatabase.cpp]: Yes, the below variables are used again 

image

in order to cover the additional references in the group by clause:

image

As all individual columns that are not part of aggregate function in the SELECT clause must also be in GROUP BY clause:

image

I hope it makes sense. I assume that the pre-processor will substitute each reference of c%02d starting from top to bottom with one variable starting from left to right and top to bottom...

Please correct my understanding if I'm wrong...

Thanks and regards

Comment (sualfred) [reply xbmc/video/VideoDatabase.cpp]: Sorry, my SQL knowledge is very limited.
I just noticed them and thought it was an issue caused by squashing commits.

I believe you if you say it's required.

Comment (wildcat2020) [reply xbmc/video/VideoDatabase.cpp]: No worries, many thanks for your checks. 
Sanity test ok. New views are now able to be queried:

image

Moving on to small scale UAT now, re-creating my video library from Trakt and running Trakt addon to see if those views still deliver the functionality Trakt expects or whether we have broken anything in terms of functionality...

This will take about an hour or so...(my library is big...)

Comment (wildcat2020) [reply xbmc/video/VideoDatabase.cpp]: Views seem to be working...

image

image

Comment (wildcat2020) [reply xbmc/video/VideoDatabase.cpp]: so scrobbling and adding data to library seems to be working fine, nothing seems to be broken so far...waiting to see trakt synchronization afterwards...

Comment (pkerling) [reply xbmc/video/VideoDatabase.cpp]: > ot: PrepareSQL works internally like printf you could be able use numbered argument to remove duplicity

I don't think that is possible. Plain printf and the custom implementation for `PrepareSQL` does not support numbering arguments.

File: xbmc/video/VideoDatabase.cpp [hunk 4]
```diff
@@ -394,12 +394,19 @@ void CVideoDatabase::CreateViews()
...
                                        "        files.idFile=episode.idFile "
-                                       "    GROUP BY tvshow.idShow");
+                                       "GROUP BY tvshow.idShow");
   m_pDS->exec(tvshowcounts);
 
   CLog::Log(LOGINFO, "create tvshow_view");
   std::string tvshowview = PrepareSQL("CREATE VIEW tvshow_view AS SELECT "
-                                     "  tvshow.*,"
+                                     "  tvshow.idShow AS idShow,"
+				     "  tvshow.c00 AS c00, tvshow.c01 AS c01, tvshow.c02 AS c02, tvshow.c03 AS c03,"
```
Comment (pkerling) [inline xbmc/video/VideoDatabase.cpp]: Here also the column number constants should be used with PrepareSQL. I know it's tedious, but it's the only clean way with the strange design that we have. You can find the list of all constants and their values in `VideoDatabase.h`.

Comment (pkerling) [reply xbmc/video/VideoDatabase.cpp]: Actually... I think you can leave `tvshow.*` here. `tvshow_view` should behave mostly like `tvshow`, so including all columns is natural and we do not have a `GROUP BY` so there is no need to spell them out explicitly. Sorry for the confusion.

Comment (wildcat2020) [reply xbmc/video/VideoDatabase.cpp]: Ok, changed back tvshow_view to SELECT tvshow.* as there is no way I can safely understand which labels to use for the tvshow.c%02d columns....
Possibly you guys need to create a wiki to document the logic of this videodatabase.h and which set of labels correspond to which tables...

**Comment by wildcat2020** [comment]
Latest build shows everything is ok...

image

File: xbmc/video/VideoDatabase.cpp (hunk 4 above)
Comment (MartijnKaijser) [reply xbmc/video/VideoDatabase.cpp]: You mean https://kodi.wiki/view/Databases/MyVideos
Database is gong to be redesigned anyway soon so if you are willing to help on that
https://github.com/xbmc/xbmc/tree/feature_odb

**Review by yol — APPROVED** [review]

**Comment by pkerling** [comment]
@MartijnKaijser If Jenkins is happy I think we can merge. Videodb behaved normally for me.

File: xbmc/video/VideoDatabase.cpp (hunk 4 above)
Comment (wildcat2020) [reply xbmc/video/VideoDatabase.cpp]: Cool, many thanks. Let me know how I can help...

**Comment by MartijnKaijser** [comment]
@wildcat2020 thx for playing along :)
Don't hesitate to send more PRs if you find something

**Comment by wildcat2020** [comment]
Great working with you guys, it's been fun. Happy to be of help. I'll let you know if I do find anything else.

File: xbmc/video/VideoDatabase.cpp (hunk 4 above)
Comment (pkerling) [reply xbmc/video/VideoDatabase.cpp]: @lobermann Feel free to ask if you have something concrete you need help with. I'm not really up-to-date on the DB rework state.

File: xbmc/video/VideoDatabase.cpp [hunk 5]
```diff
@@ -419,13 +419,17 @@ void CVideoDatabase::CreateViews()
                                      "  LEFT JOIN rating ON"
                                      "    rating.rating_id=tvshow.c%02d "
                                      "  LEFT JOIN uniqueid ON"
-                                     "    uniqueid.uniqueid_id=tvshow.c%02d "
-                                     "GROUP BY tvshow.idShow",
```
Comment (phate89) [inline xbmc/video/VideoDatabase.cpp]: I know I'm quite late but this might cause duplicated entries. The tvshowlinkpath is there actually because you might have more than one path for a TV show..this means that without grouping you'll get a duplicated TV show from the query

Comment (wildcat2020) [reply xbmc/video/VideoDatabase.cpp]: Hi, as the tvshowlinkpath is not within an aggregate function, then even with group by  (or with SELECT DISTINCT) different tvshowlinkpath values will be selected in a different row. Does it really have to be in there? Which functionality is using this view?

**Comment by phate89** [comment]
Actually with the group by you'll get only one of the paths (the first one if I'm not wrong) because only one per id.
I looked a while ago to these queries and I agree that it doesn't make a lot of sense to get one random path if you have more than one but I'm not sure why exactly is there. 
Filtering doesn't use this path because it's unreliable. I think it's used to provide a path field for skins even when you don't gather the full details (we only retrieve full information in some cases, when we load the TV show list for example we gather only the data in the view) but I have no idea if it's really used or not

**Comment by wildcat2020** [comment]
Hi, group by groups whatever is in the select clause so what you're saying is unfortunately not correct. As long as the path is in the select clause even with group by you'll get different rows

**Comment by pkerling** [comment]
@phate89 As you already looked into those queries some time ago, could you find the time to analyze the current situation again, i.e. do we have any functional changes that you can pull out with some specifically crafted demo data? I think that would save some time compared to someone new digging into it.

**Comment by wildcat2020** [comment]
I will be happy to, if you point me to the files that include the code for the required functionality

**Comment by phate89** [comment]
@wildcat2020 nope. group by reduces always to one instance in the group clause. so if one tvshow has more than one path with the group by you'll only have one row with that tv show.
it is consistent with sqlite and mysql (mysql have options to enforce full group by clause but without that it works like this):
https://rextester.com/KOINR49024
https://sqliteonline.com/#fiddle-5c308d4f7181ejmyjqjcryjp
@pkerling sure.. add in 2 separate folder the same tv show and scrape both folders.. you'll get a duplicated tv show.
a db example:
https://www.mediafire.com/file/x5941uv1800zgua/MyVideos114.db/file

**Comment by wildcat2020** [comment]
@phate89 I understand this but the problem is that this whole PR is about making these views to be able to run with full group by which is by default enabled in the latest releases of mysql.
It of course depends on the direction of Kodi, I have no saying there, but if you want it to work with other DBs than sqllite and mysql with not full group by enabled, you need to change the logic. Otherwise reject this PR and update documentation on the needed mysql setting for DB library to work.
This is your choice guys :-)

**Comment by wildcat2020** [comment]
Also,
Please consider what's  the point of having a random path feeded to kodi functionality...
In the end, It all gets down eventually to whatever kodi needs to be able to do. The views should be built in a way to support that.

**Comment by pkerling** [comment]
> mysql have options to enforce full group by clause but without that it works like this

You have to keep in mind that that option is default-enabled in newer MySQL versions.

@phate89 What I think we need is a way to write this query so that it is deterministic and still produces the results we want. As far as I understood it, this would be possible by removing the (randomly selected) path from the result? And the only thing keeping us from doing that is that we are not sure whether it is needed somewhere?

**Comment by wildcat2020** [comment]
For example: one possible use case could be the following:

Display list of TV show
User clicks on one with auto-play enabled.
Then Kodi should either start playing if only one path is available or display Play Using window and allow user to choose the appropriate source (or path if you like it)

For the above use case, you would need two views: the current one with count(TVShowstrpath)  which will give you how many paths are available
And one for listing the individual TVShowstrpath for a TV show...

I hope that makes sense

**Comment by wildcat2020** [comment]
Actually, now that I think about it, it does not make sense to have path at TV show level as each individual episode could potentially have different path :-)

**Comment by wildcat2020** [comment]
@pkerling you are absolutely correct. Like I listed in the use case I described, you would only need one line and at max you would need a count of paths available. Not each individual path itself. Yet again it makes no sense to even have the count of paths available in TV shows view, but in episodes view instead.

**Comment by pkerling** [comment]
@phate89 Found a place where it is used.

screenshot from 2019-01-05 14-32-12

Question is: Is that information helpful? Should we display all paths there instead? Or can we show the first one so it is at least deterministic?

**Comment by wildcat2020** [comment]
From my point of view, this information is useless at TV show level.
If you would like to consider something like the below functionality  (which would make sense from my point of view):

"Watch on: Netflix Vudu YouTube /data/...."

And be able to click on each source to start playing, 

You would anyway make it at episode level, not at TV Show level...

**Comment by DaVukovic** [comment]
Tbh, I mostly find this information pretty usefull. Haven't read the full story, but at least at the episode level and while at movies we should keep this info as that's the only section where you get the info about the file behind the item and if it's a mkv, an ISO or whatever. Pretty helpfull if you have different kind of media stored and also might be pretty helpful to debug problems if you know where the items comes from (smb, nfs, local)

**Comment by wildcat2020** [comment]
I agree, this information is very useful at episode level if you can action on it ofcourse (I.e.choose from which source to play from) it could be also different quality files, like 1080 720 480...
But not at TV show level.

**Comment by phate89** [comment]
@DaVukovic yes it is useful but it's potentially incorrect. You might have more than one path.. and that field only display one of the paths..
@pkerling yes that is the path but actually in the info dialog we gather the full details of a video so we might actually espose all the paths in that field if we want with some change.
@wildcat2020 yes but current form isn't viable either. And we're finalizing leia so dropping the field isn't a viable solution. I can't think a solution to restore the same exact behavior changing only the query (at least not without a couple of subqueries).. we could still use a group by and take the min value (the first in alfabetical order) or edit the tv show table to have an external id to the "main" path and use that (not a great solution imo since it requires updates in several places when we drop the path)..

for the future the ideal solution could be drop the tv show "path" fields and expose a "paths" field with all the paths

**Comment by pkerling** [comment]
> Haven't read the full story, but at least at the episode level and while at movies we should keep this info 

Sure. This is *strictly* about the TV show level.

> we could still use a group by and take the min value (the first in alfabetical order)

That sounds like the most sensible thing to do for v18. Can you cook up a PR for that?

**Comment by wildcat2020** [comment]
I could restore my branch and re-push also the change with minimum path?

**Comment by pkerling** [comment]
@wildcat2020 If you want to try that, sure. But you will have to open a new PR in any case.

**Comment by wildcat2020** [comment]
I will be lost...better somebody else do it....I am not even certain how to delete my branch at local level after I deleted it from within GIT web interface :-)

**Comment by pkerling** [comment]
`git branch -d [branch name]`

**Comment by pkerling** [comment]
You can also just push it again and it should re-appear :-) You can then open a new PR from there.

**Comment by wildcat2020** [comment]
Ok, I deleted the local branch and create a new one. 
However, now the file create is the one before this pull request.
How can I instruct my branch to take into account this pull request so that I get the most recent file?

**Comment by wildcat2020** [comment]
Ok, no worries, deleted the new one and restored the old one. Now I will try and create new pull request for the new change...

**Comment by wildcat2020** [comment]
anyway, created PR: [database] Set tvshow_view return minimum path #15200

**Comment by pkerling** [comment]
> But just realized....by introducing aggregate function to tvshow_view we now need full group by clause to be built...

Yes that cannot be avoided unfortunately.

**Comment by andbo467** [comment]
Great! Now my MySQL (MariaDB 10) database works again :)

**Comment by pkerling** [comment]
That's interesting since MariaDB does not activate `ONLY_FULL_GROUP_BY` by default, so you should not have had this problem in the first place.
